flink-connector-mqtt/src/main/java/org/apache/flink/streaming/connectors/mqtt/MqttSource.java

```diff
@@ -0,0 +1,229 @@
+
+    // Mqtt subcriber client id
+    private static final String MQTT_CLIENT_ID = MqttClient.generateClientId();
+    // Blocking queue max capacity
+    private static final int MAX_QUEUE_CAPACITY = 10000;
+
+    // Subscribe one or more topics
+    private final String[] topics;
+    // Convert bytes to output message
+    private final DeserializationSchema<OUT> deserializationSchema;
+    // User-supplied properties
+    private final Properties properties;
+    // Queue cache data form mqtt client
+    private final LinkedBlockingQueue<MqttMessage> queue;
+
+    private String serverUrl;
+    private String username;
+    private String password;
+    private String clientId;
+    private boolean cleanSession;
+    private int qos;
+
+    private transient ListState<String> mqttState;
+    private transient MqttClient client;
+
+    private volatile boolean isRunning;
+
+    public MqttSource(String topic, DeserializationSchema<OUT> deserializationSchema, Properties properties) {
+        this(Collections.singletonList(topic), deserializationSchema, properties, MAX_QUEUE_CAPACITY);
+    }
+
+    public MqttSource(List<String> topics, DeserializationSchema<OUT> deserializationSchema, Properties properties) {
+        this(topics, deserializationSchema, properties, MAX_QUEUE_CAPACITY);
+    }
+
+    /**
+     * Creates {@link MqttSource} for Streaming
+     *
+     * @param topics                Subscribe topic list
+     * @param deserializationSchema Convert bytes to output message
+     * @param properties            Mqtt properties
+     * @param capacity              Blocking queue capacity
+     */
+    public MqttSource(List<String> topics,
+                      DeserializationSchema<OUT> deserializationSchema,
+                      Properties properties,
+                      int capacity) {
+        checkProperty(properties, MqttConfig.SERVER_URL);
+        checkProperty(properties, MqttConfig.USERNAME);
+        checkProperty(properties, MqttConfig.PASSWORD);
+
+        this.topics = topics.toArray(new String[topics.size()]);
+        this.deserializationSchema = deserializationSchema;
+        this.properties = properties;
+        this.queue = new LinkedBlockingQueue<>(capacity);
+    }
+
+    @Override
+    public void initializeState(FunctionInitializationContext context) throws Exception {
+        this.mqttState = context.getOperatorStateStore().getListState(
+            new ListStateDescriptor<>("mqttClientId", String.class));
+        this.clientId = context.isRestored() ? getOnlyElement(mqttState.get()) : MQTT_CLIENT_ID;
+        LOG.info("Current mqtt subcriber clientId is: {}", clientId);
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+        if (connect()) {
+            this.isRunning = true;
+        }
+    }
+
+    private boolean connect() throws Exception {
+        this.serverUrl = properties.getProperty(MqttConfig.SERVER_URL);
```

Comment:
why we need to save this values?
count be used directly where needed?

Reply:
It's ok to use directly where needed, but this values may be used in multiple places, such as `properties.getProperty(MqttConfig.SERVER_URL)`, so saving them separately can ensure the code is concise.